csp/csp.go

```diff
@@ -260,14 +260,40 @@ func (sp *StoragePlugin) Serve(ctx context.Context, lis net.Listener) error {
 		sp.server = grpc.NewServer(sp.ServerOpts...)
 
 		// Register the CSI services.
-		if s := sp.Controller; s != nil {
-			csi.RegisterControllerServer(sp.server, s)
-		}
+		// Always register the Identity service
 		if s := sp.Identity; s != nil {
 			csi.RegisterIdentityServer(sp.server, s)
+			log.Debug("Registered Identity Service")
+		}
+
+		// determine which of the controller/node services to register
+		nsvc := true
+		csvc := true
+		if s, ok := gocsi.LookupEnv(ctx, EnvVarNodeSvcOnly); ok {
+			if b, err := strconv.ParseBool(s); err != nil {
+				if !b {
+					csvc = false
+				}
+			}
+		}
+		if s, ok := gocsi.LookupEnv(ctx, EnvVarCtrlSvcOnly); ok {
+			if b, err := strconv.ParseBool(s); err != nil {
+				if !b {
+					nsvc = false
+				}
+			}
+		}
+		if !nsvc && !csvc {
+			log.Fatalf("cannot specify both %s and %s as true",
+				EnvVarNodeSvcOnly, EnvVarCtrlSvcOnly)
+		}
+		if s := sp.Controller; s != nil && csvc {
```

Comment:
Hi @codenrhoden,

I think you should only do the env var check if the respective service is not nil. That or exit with an error if the service mode is restricted and that service is nil.

Reply:
Hi @codenrhoden,

It can be nil because the `StoragePlugin` is a struct with fields, not an interface. And the fields can be set to nil.